### PR TITLE
Update TestNoRestraint and TestWindup

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
This PR tightens up some queueset tests, both because that is good on its own and in preparation for further work on queueset and its tests.

Make TestNoRestraint verify that fairness is NOT achieved
when there is real competition.

Make TestWindup run two cases, to show that 0.1 is too narrow
a margin and 0.26 is wide enough.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by MikeSpreitzer** [comment]
@kubernetes/sig-api-machinery-misc 
/cc @wojtek-t 
/cc @tkashem 
@yue9944882

**Comment by MikeSpreitzer** [comment]
@deads2k 
@lavalamp

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go [hunk 1]
```diff
@@ -428,9 +429,9 @@ func TestNoRestraint(t *testing.T) {
 			newUniformClient(1001001001, 5, 10, time.Second, time.Second),
 			newUniformClient(2002002002, 2, 10, time.Second, time.Second/2),
 		},
-		concurrencyLimit:       10,
+		concurrencyLimit:       2,
 		evalDuration:           time.Second * 15,
-		expectedFair:           []bool{true},
+		expectedFair:           []bool{false},
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:434]: To double check - this was simply a bug that it was marked as fair, right?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go [hunk 2]
```diff
@@ -746,38 +747,45 @@ func TestTooWide(t *testing.T) {
 
 func TestWindup(t *testing.T) {
 	metrics.Register()
-	now := time.Now()
+	for _, testcase := range []struct {
+		margin2     float64
+		expectFair2 bool
+	}{{margin2: 0.26, expectFair2: true}, {margin2: 0.1, expectFair2: false}} {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:753]: nit: can we define the cases as we mostly try to do across codebase:
```
testCases := []struct {
	name        string
	margin2     float64
	expectFair2 bool
}{
  {name: "fair", margin2: 0.26, expectFair2: true},
  {name: "unfair", margin2: 0.1, expectFair2: false},
}

for _, testcase := range testCases {
```

**Review by wojtek-t — COMMENTED** [review]
Just some small nits.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go (hunk 1 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:434]: Not exactly.  With the larger concurrency limit, the NoRestraint implementation yields the fair result.  The real bug was not testing with a small concurrency limit.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go (hunk 2 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:753]: Are you concerned about having a name for the slice of test cases, or putting each case on its own line, or both?  I can imagine logic for the latter (making line-based edits independent), but the former is not obvious to me.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go (hunk 1 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:434]: "bug" is perhaps too strong a word here.  I think it is more interesting to test that the NoRestraint queueset is unfair when there is competition than it is to show that the fair result is delivered in the absence of competition.  However, there is some value in the latter --- it gives some degree of validation to test harness and its evaluation method.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go [hunk 3]
```diff
@@ -746,38 +759,48 @@ func TestTooWide(t *testing.T) {
 
 func TestWindup(t *testing.T) {
 	metrics.Register()
-	now := time.Now()
+	testCases := []struct {
+		margin2     float64
+		expectFair2 bool
+	}{
+		{margin2: 0.26, expectFair2: true},
+		{margin2: 0.1, expectFair2: false}}
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:767]: nit: for consistency with the other test:
```
   ...
   {margin2: 0.1, expectFair2: false},
}
```

Re name - I generally like having a name that already somewhat describes what the tests is supposed to test. But I can live with what you currently have too.

Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:767]: The first point here is to use comma and linebreak after the last test case, right?

I am not sure I understand the second point.  Is it about the func name, TestWindup?  This func is about a scenario that has the windup problem.

**Comment by MikeSpreitzer** [comment]
The force-push to 24cdf838ae2 adds the linebreak that I think was requested above and a comment explaining the `TestWindup` func.

**Comment by MikeSpreitzer** [comment]
The force-push to 0ee1a7b4ff9 expands that comment a little.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go [hunk 4]
```diff
@@ -744,40 +757,64 @@ func TestTooWide(t *testing.T) {
...
-	}
-	qsc, err := qsf.BeginConstruction(qCfg, newObserverPair(clk))
-	if err != nil {
-		t.Fatal(err)
+	testCases := []struct {
+		margin2     float64
+		expectFair2 bool
+	}{
+		{margin2: 0.26, expectFair2: true},
+		{margin2: 0.1, expectFair2: false},
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:780]: > I am not sure I understand the second point. Is it about the func name, TestWindup? This func is about a scenario that has the windup problem.

It's not about func name - it's about introducing an explicit "name" field in the testcase struct.
This allows a reader to quickly understand what a particular testcase is actually testing, which at least for me - even though I read those tests multiple times, I never remember what margin2 is for example.

**Pull request merged**